## Analyzing CIA Factbook Data Using SQL
In this project, I'll use SQL to explore and analyze data from the CIA World Factbook, a compendium of statistics about all of the countries on Earth. The Factbook contains demographic information like:
- population - The population as of 2015.
- population_growth - The annual population growth rate, as a percentage.
- area - The total land and water area.

In [1]:
%%capture
%load_ext sql
%sql sqlite:///factbook.db

In [3]:
%%sql
-- Get all tables in the database
SELECT * 
FROM sqlite_master
WHERE type='table';

 * sqlite:///factbook.db
Done.


[('table', 'sqlite_sequence', 'sqlite_sequence', 3, 'CREATE TABLE sqlite_sequence(name,seq)'),
 ('table', 'facts', 'facts', 47, 'CREATE TABLE "facts" ("id" INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL, "code" varchar(255) NOT NULL, "name" varchar(255) NOT NULL, "area" integer, "a ... (4 characters truncated) ... land" integer, "area_water" integer, "population" integer, "population_growth" float, "birth_rate" float, "death_rate" float, "migration_rate" float)')]

We observe there are two tables in the factbook database, namely ```sqlite_sequence```, ```facts```

Next we write a query to return first 5 rows of the facts table

In [4]:
%%sql
SELECT * 
FROM facts
LIMIT 5;

Done.


[(1, 'af', 'Afghanistan', 652230, 652230, 0, 32564342, 2.32, 38.57, 13.89, 1.51),
 (2, 'al', 'Albania', 28748, 27398, 1350, 3029278, 0.3, 12.92, 6.58, 3.3),
 (3, 'ag', 'Algeria', 2381741, 2381741, 0, 39542166, 1.84, 23.67, 4.31, 0.92),
 (4, 'an', 'Andorra', 468, 468, 0, 85580, 0.12, 8.13, 6.96, 0.0),
 (5, 'ao', 'Angola', 1246700, 1246700, 0, 19625353, 2.78, 38.78, 11.49, 0.46)]

Here are the descriptions for some of the columns:
- name - The name of the country.
- area- The country's total area (both land and water).
- area_land - The country's land area in square kilometers.
- area_water - The country's waterarea in square kilometers.
- population - The country's population.
- population_growth- The country's population growth as a percentage.
- birth_rate - The country's birth rate, or the number of births a year per 1,000 people.
- death_rate - The country's death rate, or the number of death a year per 1,000 people.

Let's start by calculating some summary statistics and look for any outlier countries.

In [5]:
%%sql
--Minimum/Maximum Population and Population_growth

SELECT  MIN(population) as min_population,
        MAX(population) as max_population,
        MIN(population_growth) as min_population_growth, 
        MAX(population_growth) as max_population_growth
FROM facts;    

Done.


[(0, 7256490011, 0.0, 4.02)]

A few things stick out from the summary statistics in the last screen:

- There's a country with a population of 0
- There's a country with a population of 7256490011 (or more than 7.2 billion people)

Let's use subqueries to zoom in on just these countries without using the specific values.

In [6]:
%%sql
--Least Populated Country

SELECT *
FROM facts
WHERE population = (SELECT MIN(population) 
                    FROM facts);

Done.


[(250, 'ay', 'Antarctica', None, 280000, None, 0, None, None, None, None)]

In [7]:
%%sql
--Most Populated Country

SELECT *
FROM facts
WHERE population == (SELECT MAX(population) 
                    FROM facts);

Done.


[(261, 'xx', 'World', None, None, None, 7256490011, 1.08, 18.6, 7.8, None)]

It seems like the table contains a row for the whole world, which explains the population of over 7.2 billion. It also seems like the table contains a row for Antarctica, which explains the population of 0. This seems to match the CIA Factbook [page for Antarctica:](https://www.cia.gov/library/publications/the-world-factbook/geos/ay.html)

Now that we know this, we should recalculate the summary statistics we calculated earlier, while excluding the row for the whole world.

In [8]:
%%sql
--Minimum/Maximum Population and Population_growth

SELECT  MIN(population) as min_population,
        MAX(population) as max_population,
        MIN(population_growth) as min_population_growth, 
        MAX(population_growth) as max_population_growth
FROM facts
WHERE name != 'World';    

Done.


[(0, 1367485388, 0.0, 4.02)]

In [9]:
%%sql
--Average Population and Area

SELECT  AVG(population) as avg_population,
        AVG(area) as avg_area
FROM facts;    

Done.


[(62094928.32231405, 555093.546184739)]

Now, we find countries that are densely populated. To do this, we'll identify countries that have:
- Above average values for population.
- Below average values for area.

In [10]:
%%sql
SELECT * 
FROM facts
WHERE population > (SELECT  AVG(population) as avg_population
                    FROM facts) 
        AND area < (SELECT AVG(area) as avg_area
                                            FROM facts) ; 

Done.


[(14, 'bg', 'Bangladesh', 148460, 130170, 18290, 168957745, 1.6, 21.14, 5.61, 0.46),
 (65, 'gm', 'Germany', 357022, 348672, 8350, 80854408, 0.17, 8.47, 11.42, 1.24),
 (85, 'ja', 'Japan', 377915, 364485, 13430, 126919659, 0.16, 7.93, 9.51, 0.0),
 (138, 'rp', 'Philippines', 300000, 298170, 1830, 100998376, 1.61, 24.27, 6.11, 2.09),
 (173, 'th', 'Thailand', 513120, 510890, 2230, 67976405, 0.34, 11.19, 7.8, 0.0),
 (185, 'uk', 'United Kingdom', 243610, 241930, 1680, 64088222, 0.54, 12.17, 9.35, 2.54),
 (192, 'vm', 'Vietnam', 331210, 310070, 21140, 94348835, 0.97, 15.96, 5.93, 0.3)]

In [11]:
%%sql
--Most Populated Country

SELECT *
FROM facts
WHERE population == (SELECT MAX(population) 
                    FROM facts 
                    WHERE name != 'World');

Done.


[(37, 'ch', 'China', 9596960, 9326410, 270550, 1367485388, 0.45, 12.49, 7.53, 0.44)]

In [12]:
%%sql
--Fastest Growing Country

SELECT *
FROM facts
WHERE population_growth == (SELECT MAX(population_growth) 
                    FROM facts 
                    WHERE name != 'World');

Done.


[(162, 'od', 'South Sudan', 644329, None, None, 12042910, 4.02, 36.91, 8.18, 11.47)]

In [13]:
%%sql
--Countries that have more water than land

SELECT name, area, area_land, area_water, 
        CAST(area_water as float)/area *100 as water_land_ratio
FROM facts
ORDER BY water_land_ratio DESC;

Done.


[('British Indian Ocean Territory', 54400, 60, 54340, 99.88970588235294),
 ('Virgin Islands', 1910, 346, 1564, 81.8848167539267),
 ('Puerto Rico', 13791, 8870, 4921, 35.6826916104706),
 ('Bahamas, The', 13880, 10010, 3870, 27.881844380403457),
 ('Guinea-Bissau', 36125, 28120, 8005, 22.15916955017301),
 ('Malawi', 118484, 94080, 24404, 20.596873839505754),
 ('Netherlands', 41543, 33893, 7650, 18.41465469513516),
 ('Uganda', 241038, 197100, 43938, 18.228661040997686),
 ('Eritrea', 117600, 101000, 16600, 14.1156462585034),
 ('Liberia', 111369, 96320, 15049, 13.512736937567905),
 ('Bangladesh', 148460, 130170, 18290, 12.319816785666173),
 ('Gambia, The', 11300, 10120, 1180, 10.442477876106194),
 ('Taiwan', 35980, 32260, 3720, 10.339077265147303),
 ('Finland', 338145, 303815, 34330, 10.152449393011874),
 ('India', 3287263, 2973193, 314070, 9.554148846624075),
 ('Ethiopia', 1104300, None, 104300, 9.444897219958346),
 ('Canada', 9984670, 9093507, 891163, 8.925312504068737),
 ('Sweden', 450295, 410335, 39960, 8.874182480373976),
 ('Colombia', 1138910, 1038700, 100210, 8.798763730233294),
 ('Brunei', 5765, 5265, 500, 8.673026886383347),
 ('Guyana', 214969, 196849, 18120, 8.429122338569748),
 ('French Polynesia', 4167, 3827, 340, 8.159347252219822),
 ('Nicaragua', 130370, 119990, 10380, 7.961954437370561),
 ('Burundi', 27830, 25680, 2150, 7.725476104922745),
 ('Iran', 1648195, 1531595, 116600, 7.074405637682435),
 ('United States', 9826675, 9161966, 664709, 6.764332798225239),
 ('Tanzania', 947300, 885800, 61500, 6.492135543122559),
 ('Vietnam', 331210, 310070, 21140, 6.382657528456266),
 ('Rwanda', 26338, 24668, 1670, 6.3406484926721856),
 ('Estonia', 45228, 42388, 2840, 6.279296011320421),
 ('Norway', 323802, 304282, 19520, 6.028375365192309),
 ('Armenia', 29743, 28203, 1540, 5.177688867968934),
 ('Tunisia', 163610, 155360, 8250, 5.042479066071756),
 ('Uzbekistan', 447400, 425400, 22000, 4.917299955297273),
 ('Indonesia', 1904569, 1811569, 93000, 4.882994525270547),
 ('Suriname', 163820, 156000, 7820, 4.773531925283849),
 ('Albania', 28748, 27398, 1350, 4.695978850702658),
 ('Ghana', 238533, 227533, 11000, 4.611521257016849),
 ('Azerbaijan', 86600, 82629, 3971, 4.585450346420323),
 ('Togo', 56785, 54385, 2400, 4.226468257462359),
 ('Russia', 17098242, 16377742, 720500, 4.2138835091935185),
 ('Kyrgyzstan', 199951, 191801, 8150, 4.075998619661817),
 ('Tonga', 747, 717, 30, 4.016064257028113),
 ('Ukraine', 603550, 579330, 24220, 4.012923535746831),
 ('Lithuania', 65300, 62680, 2620, 4.012251148545176),
 ('West Bank', 5860, 5640, 220, 3.754266211604096),
 ('Gabon', 267667, 257667, 10000, 3.7359853848251747),
 ('Turkmenistan', 488100, 469930, 18170, 3.72259782831387),
 ('Hungary', 93028, 89608, 3420, 3.6763125080620886),
 ('Latvia', 64589, 62249, 2340, 3.6229079254981498),
 ('Romania', 238391, 229891, 8500, 3.5655708478927477),
 ('Japan', 377915, 364485, 13430, 3.5537091674053687),
 ('Burma', 676578, 653508, 23070, 3.4098064081303265),
 ('Congo, Democratic Republic of the', 2344858, 2267048, 77810, 3.3183246064367227),
 ('Venezuela', 912050, 882050, 30000, 3.2892933501452775),
 ('Pakistan', 796095, 770875, 25220, 3.167963622431996),
 ('Hong Kong', 1108, 1073, 35, 3.15884476534296),
 ('Solomon Islands', 28896, 27986, 910, 3.14922480620155),
 ('Switzerland', 41277, 39997, 1280, 3.1010005572110377),
 ('Moldova', 33851, 32891, 960, 2.8359575788012172),
 ('China', 9596960, 9326410, 270550, 2.819121888598056),
 ('Korea, South', 99720, 96920, 2800, 2.807862013638187),
 ('Poland', 312685, 304255, 8430, 2.696003965652334),
 ('Iceland', 103000, 100250, 2750, 2.669902912621359),
 ('Montenegro', 13812, 13452, 360, 2.6064291920069502),
 ('Nepal', 147181, 143351, 3830, 2.6022380606192375),
 ('Botswana', 581730, 566730, 15000, 2.5785158063018927),
 ('Laos', 236800, 230800, 6000, 2.5337837837837838),
 ('Cambodia', 181035, 176515, 4520, 2.4967547711768443),
 ('Italy', 301340, 294140, 7200, 2.38932766974182),
 ('Ecuador', 283561, 276841, 6720, 2.36986045330

In [16]:
%%sql
--Fastest Growing Countries

SELECT name, population, 
        population_growth 
FROM facts
ORDER BY population_growth DESC
LIMIT 10;

Done.


[('South Sudan', 12042910, 4.02),
 ('Malawi', 17964697, 3.32),
 ('Burundi', 10742276, 3.28),
 ('Niger', 18045729, 3.25),
 ('Uganda', 37101745, 3.24),
 ('Qatar', 2194817, 3.07),
 ('Burkina Faso', 18931686, 3.03),
 ('Mali', 16955536, 2.98),
 ('Cook Islands', 9838, 2.95),
 ('Iraq', 37056169, 2.93)]

In [30]:
%%sql
--Countries adding the most people next year, 2016

SELECT name, population, 
        CAST(population_growth * population as Float)/100 as population_added 
FROM facts
WHERE name != 'World'
ORDER BY population_added DESC
LIMIT 14;

Done.


[('India', 1251695584, 15270686.1248),
 ('China', 1367485388, 6153684.246),
 ('Nigeria', 181562056, 4448270.372),
 ('Pakistan', 199085847, 2906653.3662),
 ('Ethiopia', 99465819, 2874562.1691),
 ('Bangladesh', 168957745, 2703323.92),
 ('United States', 321368864, 2506677.1392),
 ('Indonesia', 255993674, 2355141.8008000003),
 ('Congo, Democratic Republic of the', 79375136, 1944690.8320000002),
 ('Philippines', 100998376, 1626073.8536),
 ('Egypt', 88487396, 1583924.3884),
 ('Brazil', 204259812, 1572800.5524000002),
 ('Mexico', 121736809, 1436494.3462),
 ('Tanzania', 51045882, 1424180.1078)]

In [19]:
%%sql
--Countries that have higher death rate than birth rate

SELECT name, birth_rate, death_rate
FROM facts
WHERE death_rate > birth_rate
ORDER BY death_rate DESC;

Done.


[('Ukraine', 10.72, 14.46),
 ('Bulgaria', 8.92, 14.44),
 ('Latvia', 10.0, 14.31),
 ('Lithuania', 10.1, 14.27),
 ('Russia', 11.6, 13.69),
 ('Serbia', 9.08, 13.66),
 ('Belarus', 10.7, 13.36),
 ('Hungary', 9.16, 12.73),
 ('Moldova', 12.0, 12.59),
 ('Estonia', 10.51, 12.4),
 ('Croatia', 9.45, 12.18),
 ('Romania', 9.14, 11.9),
 ('Germany', 8.47, 11.42),
 ('Slovenia', 8.42, 11.37),
 ('Greece', 8.66, 11.09),
 ('Portugal', 9.27, 11.02),
 ('Czech Republic', 9.63, 10.34),
 ('Italy', 8.74, 10.19),
 ('Poland', 9.74, 10.19),
 ('Bosnia and Herzegovina', 8.87, 9.75),
 ('Saint Pierre and Miquelon', 7.42, 9.72),
 ('Japan', 7.93, 9.51),
 ('Austria', 9.41, 9.42),
 ('Monaco', 6.65, 9.24)]

In [22]:
%%sql
--Countries that have higher birth rate than death rate

SELECT name, birth_rate, death_rate
FROM facts
WHERE death_rate < birth_rate
ORDER BY death_rate DESC
LIMIT 15;

Done.


[('Lesotho', 25.47, 14.89),
 ('Guinea-Bissau', 33.38, 14.33),
 ('Chad', 36.6, 14.28),
 ('Namibia', 19.8, 13.91),
 ('Afghanistan', 38.57, 13.89),
 ('Central African Republic', 35.08, 13.8),
 ('Somalia', 40.45, 13.62),
 ('Swaziland', 24.67, 13.56),
 ('Botswana', 20.96, 13.39),
 ('Gabon', 34.49, 13.12),
 ('Nigeria', 37.64, 12.9),
 ('Mali', 44.99, 12.89),
 ('Zambia', 42.13, 12.67),
 ('Niger', 45.45, 12.42),
 ('Mozambique', 38.58, 12.1)]

In [31]:
%%sql
--Most densly populated countries

SELECT name, population, area, 
        CAST(population as float)/area *100 as population_density
FROM facts
ORDER BY population_density DESC
LIMIT 13;

Done.


[('Macau', 592731, 28, 2116896.4285714286),
 ('Monaco', 30535, 2, 1526750.0),
 ('Singapore', 5674472, 697, 814127.9770444763),
 ('Hong Kong', 7141106, 1108, 644504.1516245487),
 ('Gaza Strip', 1869055, 360, 519181.94444444444),
 ('Gibraltar', 29258, 6, 487633.3333333333),
 ('Bahrain', 1346613, 760, 177185.92105263157),
 ('Maldives', 393253, 298, 131964.09395973154),
 ('Malta', 413965, 316, 131001.58227848102),
 ('Bermuda', 70196, 54, 129992.59259259258),
 ('Sint Maarten', 39689, 34, 116732.35294117646),
 ('Bangladesh', 168957745, 148460, 113806.91432035565),
 ('Guernsey', 66080, 78, 84717.94871794872)]